### **M7T1**
Crea 4 variables, a una assigna-li la teva edat, a un altre el teu nom, a un altre el teu cognom i a un altre el valor true si estàs a l'atur o false si no estàs.

Imprimeix aquestes 4 variables anteriors juntament amb la frase "El meu nom és: X, el meu cognom és: Y i tinc Z anys. Estic a l'atur? W"" on X, Y, Z i W són els valors de les variables que has creat abans.

Demana 3 dades més mitjançant 3 promps i printa una frase amb els 3 valors que has demanat (pots inventar la temàtica de les dades que demanis).

In [20]:
X="Noelia"
Y="Solano"
Z=24
W=True
print("El meu nom és:",X,", el meu cognom és:",Y,"i tinc",Z,"anys. Estic a l'atur?",W)
ciudad=input("Introduce tu ciudad")
telefono=input("Introduce tu telefono")
mail=input("Introduce tu mail")
print("Vivo en "+ciudad,",","mi numero es "+telefono,"y mi correo es "+mail)

El meu nom és: Noelia , el meu cognom és: Solano i tinc 24 anys. Estic a l'atur? True
Vivo en  , mi numero es  y mi correo es 


### **M7T2**
Crea un programa que demani l'edat a l'usuari/ària. Si aquesta és igual o major a 18, el programa ha d'imprimir per pantalla "Ets major d'edat" si no "NO ets major d'edat."

In [4]:
edat = int(input("Quants anys tens?"))
print(edat)
if edat >= 18:
    print("Ets major d'edat")
else:
    print("NO ets major d'edat")

ValueError: invalid literal for int() with base 10: ''

### **M7T3**
Modifica el programa anterior i fes que la resposta segons l'edat segueixi aquest algoritme:

Si l'edat és negativa, el programa ha d'imprimir: "una edat negativa no és possible per a una persona."

Si l'edat està entre 0 i 17 (ambdós inclosos), el programa ha d'imprimir: "ets MENOR d'edat."

Si l'edat està entre 18 i 125 (ambdós inclosos), el programa ha d'imprimir: "ets MAJOR d'edat."

Si l'edat és major que 125, el programa ha d'imprimir: "una edat tan alta és difícil de veure en una persona."

In [3]:
edat = int(input("Quants anys tens?"))
print(edat)
if edat < 0:
    print("una edat negativa no és possible per a una persona.")
elif edat >= 0:
    print("ets MENOR d'edat.")
elif edat >= 18:
    print("ets major d'edat")
elif edat > 125:
    print("una edat tan alta és difícil de veure en una persona.")

125
ets MENOR d'edat.


### **M7T4**
Crea una funció que demani el nom a l'usuari i mostri un missatge de salutació.

In [ ]:
def saludo(nombre):
    print(f"Bienvenido/a {nombre}. ¿Cómo te encuentras hoy?")
saludo("Noelia")

Bienvenido/a Noelia. ¿Cómo te encuentras hoy?


In [18]:
def saludo(nombre):
    print(f"Bienvenido/a {nombre}. ¿Cómo te encuentras hoy?")
saludo(input("Indica tu nombre"))

Bienvenido/a noelia. ¿Cómo te encuentras hoy?


### **M7T5**
Crea una funció que rebi un preu i retorni el preu amb IVA (21% per defecte).

A continuació, demana un preu a l’usuari i mostra el resultat amb l’IVA aplicat.

In [ ]:
def total(precio):
    impuesto=float(precio)*0.21
    suma=float(precio)+impuesto
    return suma
total(input("Indica el precio: "))

24.2

### **M7T6**
Crea una funció que rebi un número enter com a argument i retorni True si el número és parell, o False si és senar.

A continuació, demana un número a l'usuari, utilitza la funció i mostra un missatge indicant si el número és parell o senar.

In [ ]:
def prueba(num):
    print(int(num/2)==num/2)
    return int(num/2)==num/2
es_par=prueba(int(input("Introduce un numero entero: ")))
if es_par:
    print("El numero es par")
else:
    print("El numero es impar")

False
El numero es impar


In [2]:
def par_impar (num):
    int(num)
    if num % 2 == 0:
        return True
    else:
        return False
 
def mensaje():
    num = int(input("Ingrese un número entero"))
    check = par_impar(num)
    if check == True:
        print("El número es par")
    else:
        print("El número es impar") 
 
mensaje()

El número es par


### **M7T7**
Crea una llista amb tres tasques inicials: "netejar la cuina", "treure el gos a passejar" i "portar a la nena al col·legi". Un cop creada, imprimeix-la.

Després afegeix al final de la llista, les tasques: "fer esport", "meditar". Torna a imprimir-la per veure que s'han afegit.

Elimina la tasca "netejar la cuina" i torna a imprimir la llista.

Finalment, afegeix la tasca "cuinar" a la segona posició de la llista.

In [6]:
llista=list(("netejar la cuina", "treure el gos a passejar", "portar a la nena al col·legi"))
print(llista)
llista.extend(["fer esport","meditar"])
print(llista)
llista.remove("netejar la cuina")
print(llista)
llista.insert(1,"cuinar")
print(llista)

['netejar la cuina', 'treure el gos a passejar', 'portar a la nena al col·legi']
['netejar la cuina', 'treure el gos a passejar', 'portar a la nena al col·legi', 'fer esport', 'meditar']
['treure el gos a passejar', 'portar a la nena al col·legi', 'fer esport', 'meditar']
['treure el gos a passejar', 'cuinar', 'portar a la nena al col·legi', 'fer esport', 'meditar']


--------------------------------------------------------------------------------------------------

Crea un diccionari on puguem determinar quin dia es fa cada tasca. Assigna un dia diferent de la setmana a totes les tasques anteriors.

Printa la tasca del dilluns i donat que hi havia 6 tasques, afegeix una setena al dia de la setmana que et falti.

In [40]:
diccionari=dict(dilluns="treure el gos a passejar", dimarts="cuinar", dimecres="portar la nena al col·legi", dijous="fer esport", divendres="meditar", dissabte="netejar la cuina")
print(diccionari["dilluns"])

diccionari["diumenge"]="fer bugada"
print(diccionari)

treure el gos a passejar
{'dilluns': 'treure el gos a passejar', 'dimarts': 'cuinar', 'dimecres': 'portar la nena al col·legi', 'dijous': 'fer esport', 'divendres': 'meditar', 'dissabte': 'netejar la cuina', 'diumenge': 'fer bugada'}


### **M7T8**
Crea una funció que faci el següent:

Creï una llista buida.
Demani a l’usuari 3 paraules, una per una.
Si la paraula comença per 'A' o 'a' i té menys de 7 lletres, afegeix-la a la llista.
Si no compleix aquestes condicions, no facis res.
Un cop demanades les 3 paraules i avaluades, imprimeix la llista completa.

Finalment, crida la funció perquè s’executi.

In [27]:
def funcio():
    llista=[]
    x=input("Escriu una paraula")
    y=input("Escriu una paraula")
    z=input("Escriu una paraula")
    if (x.startswith("A") or x.startswith("a")) and len(x)<7:
        llista.append(x)
    if (y.startswith("A") or y.startswith("a")) and len(y)<7:
        llista.append(y)
    if (z.startswith("A") or z.startswith("a")) and len(z)<7:
        llista.append(z)
    return llista
print(funcio())

['aaa', 'a']


In [31]:
def definir_llista():
    llista = []
 
    paraula1 = input("Introdueix la primera paraula: ")
    if paraula1 == 'A' or 'a' and len(paraula1) < 7:
        llista.append(paraula1)
    else:
        print("Paraula incorrecta, no compleix les condicions.")
 
    paraula2 = input("Introdueix la segona paraula: ")
    if paraula2 == 'A' or 'a' and len(paraula2) < 7:
        llista.append(paraula2)
    else:
        print("Paraula incorrecta, no compleix les condicions.")
 
    paraula3 = input("Introdueix la tercera paraula: ")
    if paraula3 == 'A' or 'a' and len(paraula3) < 7:
        llista.append(paraula3)
    else:
        print("Paraula incorrecta, no compleix les condicions.")
 
    print("Llista resultant:", llista)
   
definir_llista()

Llista resultant: ['', '', '']


### **M7T9**
Crea una funció que rebi com a paràmetre el nombre de paraules que l’usuari/ària vol introduir. A continuació:

Demana tantes paraules com hagi indicat l’usuari/ària i desa-les en una llista.
Mostra per pantalla totes les paraules introduïdes.
Finalment, abans de cridar la funció, demana a l’usuari/ària quantes paraules vol introduir.

In [55]:
def funcion(numero_palabras):
    lista=[]
    for x in range(numero_palabras):
        palabras=input("Introduce palabras: ")
        lista.append(palabras)
        print(palabras)
    return lista
numero_palabras=int(input("Introduce un numero de palabras: "))
funcion(numero_palabras)

si
no
sino


['si', 'no', 'sino']

### **M7T10**
Defineix una funció que rebi una llista de números i retorni el sumatori de tots els elements (sense utilitzar sum(), fes el càlcul manualment amb un bucle).

Crea una llista amb 10 números.
Calcula la mitjana fent servir la funció del sumatori.
Mostra per pantalla el sumatori i la mitjana.

In [11]:
def funcion(lista):
    suma=0
    for x in lista:
        suma+=x #*= esto es un contador
    return suma
lista=[0,1,2,3,4,5,6,7,8,9]
sumatorio=funcion(lista)
print(f"Este es el sumatorio: {sumatorio}")
media=sumatorio/len(lista)
print(f"Esta es la media: {mediana}")


Este es el sumatorio: 45
Esta es la media: 4.5


In [2]:
def funcion(lista):
    suma=0
    for x in lista:
        suma=suma+x
    return suma
lista=[0,1,2,3,4,5,6,7,8,9]
sumatorio=funcion(lista)
print(f"Este es el sumatorio: {sumatorio}")
mediana=sumatorio/len(lista)
print(f"Esta es la media: {mediana}")

Este es el sumatorio: 45
Esta es la media: 4.5


### **M7T11**
Demani a l’usuari/ària 5 articles i el seu preu corresponent.
Desi aquests articles i preus en un diccionari, on la clau sigui el nom de l’article i el valor, el seu preu.
Retorni el diccionari amb les dades introduïdes.
Un cop creat el diccionari, imprimeix només els articles que costin menys de 10 euros.

In [7]:
def funcion():
    diccionario={}
    for x in range(5):
        articulo=input("Ingrese el nombre del artículo")
        precio=int(input(f"Ingrese el precio de {articulo}"))
        diccionario[articulo]=precio
    return diccionario
coleccion=funcion()
print (coleccion)
lista=[]
for x, y in coleccion.items():
    if y<10:
        lista.append(x)
print(f"Estos son los articulos con un precio menor de 10: {lista}")

{'hola': 2, 's': 2, 't': 55, 'r': 44, 'd': 2}
Estos son los articulos con un precio menor de 10: ['hola', 's', 'd']


In [5]:
def cinco_art():
    cincoDiccionario = {}
    for x in range(5):
        articulo = input("Ingrese el nombre del artículo")
        precioArt = int(input(f"Ingrese el precio de {articulo}"))
        cincoDiccionario[articulo] = precioArt
    return cincoDiccionario
 
def menos10(diccionario):
   for x in diccionario:
      if diccionario[x] < 10:
          print(x, diccionario[x])
 
diccionario5 = cinco_art()
 
print(diccionario5)
menos10(diccionario5)

{'ns': 55, 't': 4, 'r': 1}
t 4
r 1


In [ ]:
#otra opción
def articles_preus():
    dicc_articles = {}
    for i in range(5):
        nom_article = str(input(f"Digues el nom de l'article {1 + i}/5"))
        preu_article = int(input(f"Quin és el preu de l'article {nom_article}?"))
        dicc_articles[nom_article] = preu_article
    return dicc_articles
 
dicc_articles = articles_preus()
 
menor_10 = []
for nom in dicc_articles:
    if int(dicc_articles[nom]) < 10:
        menor_10.append(nom)
    else:
        continue
print("Articles amb preu menor a 10 euros: ")
print(*menor_10, sep = ', ')

ValueError: invalid literal for int() with base 10: ''

### **M7T12**
En aquest exercici, treballaràs amb el dataset Salut_i_son.csv utilitzant la biblioteca Pandas de Python. Realitzaràs diverses tasques d’exploració, filtres, agrupacions i modificació de columnes.

1. Exploració de dades
Carrega el dataset Salut_i_son.csv i respon les següents preguntes:

Quantes columnes i quants registres té el dataset?
Quantes columnes són de tipus enter? I quantes de tipus decimal?
Quina és l’edat màxima registrada?
Quina és la duració mínima de la son?
Quina és la mitjana de la qualitat de la son?
Explica com has trobat cadascuna de les respostes

In [25]:
import pandas as pd
df = pd.read_csv('csvs/Salut_i_son.csv')
df

Person ID  Gender  Age            Occupation  Sleep Duration  \
0            1    Male   27     Software Engineer             6.1   
1            2    Male   28                Doctor             6.2   
2            3    Male   28                Doctor             6.2   
3            4    Male   28  Sales Representative             5.9   
4            5    Male   28  Sales Representative             5.9   
..         ...     ...  ...                   ...             ...   
369        370  Female   59                 Nurse             8.1   
370        371  Female   59                 Nurse             8.0   
371        372  Female   59                 Nurse             8.1   
372        373  Female   59                 Nurse             8.1   
373        374  Female   59                 Nurse             8.1   

     Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                   6                       42             6   Overweight   
1                   6                       60             8       Normal   
2                   6                       60             8       Normal   
3                   4                       30             8        Obese   
4                   4                       30             8        Obese   
..                ...                      ...           ...          ...   
369                 9                       75             3   Overweight   
370                 9                       75             3   Overweight   
371                 9                       75             3   Overweight   
372                 9                       75             3   Overweight   
373                 9                       75             3   Overweight   

    Blood Pressure  Heart Rate  Daily Steps Sleep Disorder  
0           126/83          77         4200            NaN  
1           125/80          75        10000            NaN  
2           125/80          75        10000            NaN  
3           140/90          85         3000    Sleep Apnea  
4           140/90          85         3000    Sleep Apnea  
..             ...         ...          ...            ...  
369         140/95          68         7000    Sleep Apnea  
370         140/95          68         7000    Sleep Apnea  
371         140/95          68         7000    Sleep Apnea  
372         140/95          68         7000    Sleep Apnea  
373         140/95          68         7000    Sleep Apnea  

[374 rows x 13 columns]

1. **RESPUESTA:** El dataframe contiene 374 filas y 13 columnas. Hay una columna de tipo decimal y 7 de tipo entero La edad maxima registrada es 59 y el sueño minimo es 5,8. La media de la calidad de sueño es 7,31 Filas y columnas lo he encontrado con df.shape, el tipo de variables decimal y entero lo he encontrado con df.info() y los demás con df.describe().

In [33]:
print(f"El dataframe contiene este numero de filas y columnas respectivamente: {df.shape}")
df.info()
df.describe()

El dataframe contiene este numero de filas y columnas respectivamente: (374, 13)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 374 entries, 0 to 373
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Person ID                374 non-null    int64  
 1   Gender                   374 non-null    object 
 2   Age                      374 non-null    int64  
 3   Occupation               374 non-null    object 
 4   Sleep Duration           374 non-null    float64
 5   Quality of Sleep         374 non-null    int64  
 6   Physical Activity Level  374 non-null    int64  
 7   Stress Level             374 non-null    int64  
 8   BMI Category             374 non-null    object 
 9   Blood Pressure           374 non-null    object 
 10  Heart Rate               374 non-null    int64  
 11  Daily Steps              374 non-null    int64  
 12  Sleep Disorder           155 non-null    object 
dtyp

Person ID         Age  Sleep Duration  Quality of Sleep  \
count  374.000000  374.000000      374.000000        374.000000   
mean   187.500000   42.184492        7.132086          7.312834   
std    108.108742    8.673133        0.795657          1.196956   
min      1.000000   27.000000        5.800000          4.000000   
25%     94.250000   35.250000        6.400000          6.000000   
50%    187.500000   43.000000        7.200000          7.000000   
75%    280.750000   50.000000        7.800000          8.000000   
max    374.000000   59.000000        8.500000          9.000000   

       Physical Activity Level  Stress Level  Heart Rate   Daily Steps  
count               374.000000    374.000000  374.000000    374.000000  
mean                 59.171123      5.385027   70.165775   6816.844920  
std                  20.830804      1.774526    4.135676   1617.915679  
min                  30.000000      3.000000   65.000000   3000.000000  
25%                  45.000000      4.000000   68.000000   5600.000000  
50%                  60.000000      5.000000   70.000000   7000.000000  
75%                  75.000000      7.000000   72.000000   8000.000000  
max                  90.000000      8.000000   86.000000  10000.000000

In [ ]:
#corrección
tipos_datos = df.dtypes
enteros = tipos_datos[tipos_datos == 'int64'].count()
decimales = tipos_datos[tipos_datos == 'float64'].count()
enteros

np.int64(7)

In [ ]:
#correcion
df['Age'].max()

np.int64(59)

In [ ]:
#correcion
df['Sleep Duration'].min()

np.float64(5.8)

In [ ]:
#correcion
df['Quality of Sleep'].mean()

np.float64(7.31283422459893)

2. Filtres
Classifica les dades en dos grups de persones segons els criteris següents:

Grupo 1: Persones de més de 40 anys, amb una qualitat de la son superior a 8 i que fan més de 5000 passes al dia.

In [36]:
grupo1=df[(df['Age']>40)
   &(df['Quality of Sleep']>8)
   &(df['Daily Steps']>5000)]
print(len(grupo1))

33


Grupo 2: Persones que treballen com a enginyers/eres o enginyers/eres de software i tenen una heart rate superior a 60.

In [37]:
grupo2=df[(df['Occupation']=='Software Engineer')
   | (df['Occupation']=='Engineer')
   & (df['Heart Rate']>60)]
print(len(grupo2))

67


In [35]:
# Grupo 1:
grupo1 = df[(df['Age']>40)
   &(df['Quality of Sleep']>8)
   &(df['Daily Steps']>5000)]
# Grupo 2: Ocupación relacionada con ingeniería y heart rate > 60
grupo2 = df[(df['Occupation'].str.contains('Engineer', case=False)) &
            (df['Heart Rate'] > 60)]
print("Cantidad en el grupo 1: ", len(grupo1))
print("Cantidad en el grupo 2: ", len(grupo2))

Cantidad en el grupo 1:  33
Cantidad en el grupo 2:  67


3. Agrupacions
Agrupa les dades del dataset original per calcular valors mitjans:

Calcula la mitjana de la duració de la son agrupada per gènere.
Calcula la mitjana de la duració de la son agrupada per gènere i ocupació.

In [45]:
df.groupby('Gender')[['Sleep Duration']].mean()

Sleep Duration
Gender                
Female        7.229730
Male          7.036508

In [50]:
df.groupby(['Gender','Occupation'])[['Sleep Duration']].mean()

Sleep Duration
Gender Occupation                          
Female Accountant                  7.111111
       Doctor                      8.200000
       Engineer                    8.425000
       Lawyer                      7.150000
       Manager                     6.900000
       Nurse                       7.063014
       Scientist                   6.000000
       Teacher                     6.705714
Male   Accountant                  7.200000
       Doctor                      6.934783
       Engineer                    7.535484
       Lawyer                      7.422222
       Sales Representative        5.900000
       Salesperson                 6.403125
       Software Engineer           6.750000
       Teacher                     6.580000

4. Modificació de columnes
Afegeix una nova columna al DataFrame, anomenada "Poca Son", amb el valor True si la persona dorm menys de 7 hores, i False en cas contrari.
Elimina les últimes 4 columnes del DataFrame original i desa el nou DataFrame en una variable anomenada sleep_info.

In [29]:
df.loc[df['Sleep Duration']<7, 'Little Sleep'] = True
df.loc[df['Sleep Duration']>=7, 'Little Sleep'] = False
df

Person ID  Gender  Age            Occupation  Sleep Duration  \
0            1    Male   27     Software Engineer             6.1   
1            2    Male   28                Doctor             6.2   
2            3    Male   28                Doctor             6.2   
3            4    Male   28  Sales Representative             5.9   
4            5    Male   28  Sales Representative             5.9   
..         ...     ...  ...                   ...             ...   
369        370  Female   59                 Nurse             8.1   
370        371  Female   59                 Nurse             8.0   
371        372  Female   59                 Nurse             8.1   
372        373  Female   59                 Nurse             8.1   
373        374  Female   59                 Nurse             8.1   

     Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                   6                       42             6   Overweight   
1                   6                       60             8       Normal   
2                   6                       60             8       Normal   
3                   4                       30             8        Obese   
4                   4                       30             8        Obese   
..                ...                      ...           ...          ...   
369                 9                       75             3   Overweight   
370                 9                       75             3   Overweight   
371                 9                       75             3   Overweight   
372                 9                       75             3   Overweight   
373                 9                       75             3   Overweight   

    Blood Pressure  Heart Rate  Daily Steps Sleep Disorder Little Sleep  
0           126/83          77         4200            NaN         True  
1           125/80          75        10000            NaN         True  
2           125/80          75        10000            NaN         True  
3           140/90          85         3000    Sleep Apnea         True  
4           140/90          85         3000    Sleep Apnea         True  
..             ...         ...          ...            ...          ...  
369         140/95          68         7000    Sleep Apnea        False  
370         140/95          68         7000    Sleep Apnea        False  
371         140/95          68         7000    Sleep Apnea        False  
372         140/95          68         7000    Sleep Apnea        False  
373         140/95          68         7000    Sleep Apnea        False  

[374 rows x 14 columns]

In [ ]:
#correcion
df['Little Sleep']=df['Sleep Duration'] < 7
df

Person ID  Gender  Age            Occupation  Sleep Duration  \
0            1    Male   27     Software Engineer             6.1   
1            2    Male   28                Doctor             6.2   
2            3    Male   28                Doctor             6.2   
3            4    Male   28  Sales Representative             5.9   
4            5    Male   28  Sales Representative             5.9   
..         ...     ...  ...                   ...             ...   
369        370  Female   59                 Nurse             8.1   
370        371  Female   59                 Nurse             8.0   
371        372  Female   59                 Nurse             8.1   
372        373  Female   59                 Nurse             8.1   
373        374  Female   59                 Nurse             8.1   

     Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                   6                       42             6   Overweight   
1                   6                       60             8       Normal   
2                   6                       60             8       Normal   
3                   4                       30             8        Obese   
4                   4                       30             8        Obese   
..                ...                      ...           ...          ...   
369                 9                       75             3   Overweight   
370                 9                       75             3   Overweight   
371                 9                       75             3   Overweight   
372                 9                       75             3   Overweight   
373                 9                       75             3   Overweight   

    Blood Pressure  Heart Rate  Daily Steps Sleep Disorder  Little Sleep  
0           126/83          77         4200            NaN          True  
1           125/80          75        10000            NaN          True  
2           125/80          75        10000            NaN          True  
3           140/90          85         3000    Sleep Apnea          True  
4           140/90          85         3000    Sleep Apnea          True  
..             ...         ...          ...            ...           ...  
369         140/95          68         7000    Sleep Apnea         False  
370         140/95          68         7000    Sleep Apnea         False  
371         140/95          68         7000    Sleep Apnea         False  
372         140/95          68         7000    Sleep Apnea         False  
373         140/95          68         7000    Sleep Apnea         False  

[374 rows x 14 columns]

In [ ]:
sleep_info=df.drop(['Blood Pressure', 'Heart Rate', 'Daily Steps', 'Sleep Disorder'], axis=1).head()
sleep_info

Person ID Gender  Age            Occupation  Sleep Duration  \
0          1   Male   27     Software Engineer             6.1   
1          2   Male   28                Doctor             6.2   
2          3   Male   28                Doctor             6.2   
3          4   Male   28  Sales Representative             5.9   
4          5   Male   28  Sales Representative             5.9   

   Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                 6                       42             6   Overweight   
1                 6                       60             8       Normal   
2                 6                       60             8       Normal   
3                 4                       30             8        Obese   
4                 4                       30             8        Obese   

  Little Sleep  
0         True  
1         True  
2         True  
3         True  
4         True

In [ ]:
##correcion
sleep_info=df.iloc[:,:-4]
sleep_info

Person ID  Gender  Age            Occupation  Sleep Duration  \
0            1    Male   27     Software Engineer             6.1   
1            2    Male   28                Doctor             6.2   
2            3    Male   28                Doctor             6.2   
3            4    Male   28  Sales Representative             5.9   
4            5    Male   28  Sales Representative             5.9   
..         ...     ...  ...                   ...             ...   
369        370  Female   59                 Nurse             8.1   
370        371  Female   59                 Nurse             8.0   
371        372  Female   59                 Nurse             8.1   
372        373  Female   59                 Nurse             8.1   
373        374  Female   59                 Nurse             8.1   

     Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                   6                       42             6   Overweight   
1                   6                       60             8       Normal   
2                   6                       60             8       Normal   
3                   4                       30             8        Obese   
4                   4                       30             8        Obese   
..                ...                      ...           ...          ...   
369                 9                       75             3   Overweight   
370                 9                       75             3   Overweight   
371                 9                       75             3   Overweight   
372                 9                       75             3   Overweight   
373                 9                       75             3   Overweight   

    Blood Pressure  
0           126/83  
1           125/80  
2           125/80  
3           140/90  
4           140/90  
..             ...  
369         140/95  
370         140/95  
371         140/95  
372         140/95  
373         140/95  

[374 rows x 10 columns]

In [ ]:
##os dejo las formas que yo encontre para eliminar columnas 1. Eliminar una sola columna:
df = df.drop('NombreColumna', axis=1)
 
2. . Eliminar varias columnas:
df = df.drop(['Columna1', 'Columna2'], axis=1)
 
3. Eliminar columnas in place (sin reasignar el DataFrame):
df.drop('Columna1', axis=1, inplace=True)
 
ex
 
import pandas as pd
 
df = pd.DataFrame({
    'Nom': ['Anna', 'Joan', 'Marc'],
    'Edat': [23, 35, 30],
    'Ciutat': ['Barcelona', 'Girona', 'Tarragona']
})
 
# Eliminar la columna 'Ciutat'
df = df.drop('Ciutat', axis=1)
 
print(df)
 
Nota = axis=1 indica que quieres eliminar columnas (si pusieras axis=0, eliminarías filas).
 
4. Eliminar con la df.iloc[fila, columna]
.iloc significa index location y se usa para acceder a datos por posición numérica (índices), no por nombres de columnas o filas.
fila: índice o rango de filas (por posición, empezando en 0).
 
columna: índice o rango de columnas (también por posición).
Varias filas
df.iloc[0:3]       # Devuelve las filas 0, 1 y 2 (el final no se incluye)
una celda especifica
df.iloc[1, 2]      # Fila 1, columna 2 (índices basados en posición)
Todas las filas, una columna
df.iloc[:, 1]      # Todas las filas, columna 1
 Todas las filas, todas menos las últimas 4 columnas
 df.iloc[:, :-4]    # Todas las filas, columnas desde el inicio hasta 4 antes del final

### **M7T13**
La nostra empresa es dedica a realitzar anàlisis de dades geogràfiques i socials, i un client ens ha sol·licitat una eina per identificar dades bàsiques de cada continent de manera automàtica.

Per això, cal crear una funció que rebi dos paràmetres:

Un DataFrame amb informació de països (com el del fitxer Country.xlsx).
El nom d’un continent (com a cadena de text).

La funció ha de retornar un diccionari amb les següents dades relacionades amb el continent indicat:

Nombre de regions diferents que hi ha dins del continent.
Nombre de països dins del continent.
Superfície màxima (SurfaceArea) registrada al continent.
Superfície mínima registrada.
Població mitjana (Population) dels països del continent.
Mitjana de l’esperança de vida (LifeExpectancy) dels països del continent.

Finalment, carrega el fitxer adjunt amb la següent instrucció:

data = pd.read_csv("arxiu csv", sep=";", encoding="latin-1")

Executa la funció per a diversos continents i mostra els resultats obtinguts.

In [4]:
import pandas as pd
data = pd.read_csv("csvs/Country(in).csv", sep=";", encoding="latin-1")
data

Code          Name      Continent                     Region  SurfaceArea  \
0    ABW         Aruba  North America                  Caribbean        193.0   
1    AFG   Afghanistan           Asia  Southern and Central Asia     652090.0   
2    AGO        Angola         Africa             Central Africa    1246700.0   
3    AIA      Anguilla  North America                  Caribbean         96.0   
4    ALB       Albania         Europe            Southern Europe      28748.0   
..   ...           ...            ...                        ...          ...   
234  YEM         Yemen           Asia                Middle East     527968.0   
235  YUG    Yugoslavia         Europe            Southern Europe     102173.0   
236  ZAF  South Africa         Africa            Southern Africa    1221037.0   
237  ZMB        Zambia         Africa             Eastern Africa     752618.0   
238  ZWE      Zimbabwe         Africa             Eastern Africa     390757.0   

     IndepYear  Population  LifeExpectancy       GNP    GNPOld  \
0          NaN      103000            78.4     828.0     793.0   
1       1919.0    22720000            45.9    5976.0       NaN   
2       1975.0    12878000            38.3    6648.0    7984.0   
3          NaN        8000            76.1      63.2       NaN   
4       1912.0     3401200            71.6    3205.0    2500.0   
..         ...         ...             ...       ...       ...   
234     1918.0    18112000            59.8    6041.0    5729.0   
235     1918.0    10640000            72.4   17000.0       NaN   
236     1910.0    40377000            51.1  116729.0  129092.0   
237     1964.0     9169000            37.2    3377.0    3922.0   
238     1980.0    11669000            37.8    5951.0    8670.0   

                 LocalName                                GovernmentForm  \
0                    Aruba  Nonmetropolitan Territory of The Netherlands   
1    Afganistan/Afqanestan                               Islamic Emirate   
2                   Angola                                      Republic   
3                 Anguilla                 Dependent Territory of the UK   
4                Shqipëria                                      Republic   
..                     ...                                           ...   
234               Al-Yaman                                      Republic   
235            Jugoslavija                              Federal Republic   
236           South Africa                                      Republic   
237                 Zambia                                      Republic   
238               Zimbabwe                                      Republic   

                 HeadOfState  Capital Code2,  
0                    Beatrix    129.0    AW,  
1              Mohammad Omar      1.0    AF,  
2    José Eduardo dos Santos     56.0    AO,  
3               Elisabeth II     62.0    AI,  
4             Rexhep Mejdani     34.0    AL,  
..                       ...      ...    ...  
234       Ali Abdallah Salih   1780.0    YE,  
235       Vojislav Koštunica   1792.0    YU,  
236              Thabo Mbeki    716.0    ZA,  
237        Frederick Chiluba   3162.0    ZM,  
238         Robert G. Mugabe   4068.0    ZW,  

[239 rows x 15 columns]

In [4]:
def funcion(data, continentes):
    datac = data[data['Continent'] == continentes]
    suma_regiones=datac['Region'].nunique()
    suma_paises=datac['Name'].nunique()
    max_surface=datac['SurfaceArea'].max()
    min_surface=datac['SurfaceArea'].min()
    promedio_population=datac['Population'].mean()
    promedio_vida=datac['LifeExpectancy'].mean()
    data_continents=pd.DataFrame({'Numero de regiones': [suma_regiones], 'Numero de paises': [suma_paises], 'Area maxima': [max_surface], 'Area minima': [min_surface], 'Poblacion promedio': [promedio_population], 'Esperanza de vida promedio': [promedio_vida]})
    return data_continents
cont= sorted(data['Continent'].unique())
for continente in cont:
    print(f'\nEsta es la información del {continente}:\n', funcion(data, continente))


Esta es la información del Africa:
    Numero de regiones  Numero de paises  Area maxima  Area minima  \
0                   5                58    2505813.0         78.0   

   Poblacion promedio  Esperanza de vida promedio  
0        1.352543e+07                    52.57193  

Esta es la información del Antarctica:
    Numero de regiones  Numero de paises  Area maxima  Area minima  \
0                   1                 5   13120000.0         59.0   

   Poblacion promedio  Esperanza de vida promedio  
0                 0.0                         NaN  

Esta es la información del Asia:
    Numero de regiones  Numero de paises  Area maxima  Area minima  \
0                   4                51    9572900.0         18.0   

   Poblacion promedio  Esperanza de vida promedio  
0        7.264756e+07                   67.441176  

Esta es la información del Europe:
    Numero de regiones  Numero de paises  Area maxima  Area minima  \
0                   6                46   17075400.0

### **M7T14**
La nostra empresa està col·laborant amb una ONG que estudia la distribució de l'espai i dels recursos geogràfics en relació amb la població mundial. L’objectiu és identificar quins països tenen menys espai disponible per habitant, fet que pot estar relacionat amb situacions de sobrepoblació, pressió urbana o desigualtat en l’ús del territori.

Per dur a terme aquest estudi, ens han proporcionat un conjunt de dades (Country.csv) que conté informació detallada sobre la població, la superfície i altres atributs de cada país.


Crea una funció en Python que rebi com a paràmetres:

Un DataFrame amb dades de països.
El nombre n de països que es volen obtenir com a resultat.

La funció ha de calcular, per a cada país, la quantitat de metres quadrats disponibles per persona, utilitzant la següent fórmula:

Metres per persona = SurfaceArea × 1.000.000 / Population

La funció ha de retornar els n països amb menys espai disponible per habitant.

Executa la funció per obtenir els 10 països amb menys espai per persona i després per obtenir-ne 20.

In [ ]:
data_copia=data.copy()
def funcion(data, npaises):
    data['Metros por persona']= round((data['SurfaceArea']*1000000)/data['Population'], 2)
    ##data = data.dropna(subset=['Metros por persona']) ==> para quitar nulos, no es necesario en fundamentos
    data_paises= data.sort_values(by='Metros por persona')
    return data_paises[['Name', 'Metros por persona']].head(npaises)
funcion(data, int(input('Introduce un numero de paises: ')))

Name  Metros por persona
129      Macao               38.05
131     Monaco               44.12
92   Hong Kong              158.51
186  Singapore              173.25
79   Gibraltar              240.00

### **Per practicar més**
#### Exercicis de condicionals, funcions i bucles
Crea una funció que, donada una frase o una paraula, retorni el nombre de vocals. Un cop definida, sol·licita una paraula o frase a l’usuari i imprimeix el nombre de vocals.

In [15]:
def funcion(respuesta):
    acumulador=0
    for x in respuesta:
        if x.isalpha():
            acumulador+=1
    return acumulador
funcion(input('Ingresa una palabra o frase: '))

8


Defineix una funció que, donat un número, imprimeixi la seva taula de multiplicar. (Exemple: 1 × 1 = 1, 1 × 2 = 2, 1 × 3 = 3, ...).

In [22]:
def funcion(numero):
    tabla=[1,2,3,4,5,6,7,8,9,10]
    lista=[]
    for x in tabla:
        producto=numero*x
        lista.append(producto)
    return lista
funcion(int(input('Ingresa un numero para visualizar su tabla de multiplicar: ')))


[9, 18, 27, 36, 45, 54, 63, 72, 81, 90]

In [ ]:
def taula_multiplicar(numero):
    for i in range(1, 11):
        resultat = numero * i
        print(f"{numero} x {i} = {resultat}")

entrada = int(input("Introdueix un número per veure la seva taula de multiplicar: "))
taula_multiplicar(entrada)

Defineix una funció que, donada una frase, retorni el nombre de paraules úniques que conté, sense fer servir la funció len(). (Pots utilitzar el mètode split() d’un string.)

In [ ]:
def funcion(frase):
    palabras=frase.lower().split()
    unicas=set(palabras)
    acumulador=0
    for x in unicas:
        acumulador+=1
    return acumulador
funcion('Solo los valores unicos no los repetidos')



6

Defineix una funció que, donada una llista de números, imprimeixi només els números parells. Genera una llista d’exemple i crida la funció.

In [38]:
def funcion(lista):
    lista_par=[]
    for x in lista:
        if x % 2 == 0:
            lista_par.append(x)
        else:
            continue
    return lista_par
ejemplo=[3,2,7,8,9,10,1,33,6,2,44]
funcion(ejemplo)

[2, 8, 10, 6, 2, 44]

Escriu una funció que demani un text i retorni un diccionari amb les lletres que apareixen i la quantitat de cops que apareixen, sense fer servir funcions d’agregació com count().

In [ ]:
def funcion():
    texto=input('Añade una frase de texto para contar cada letra y su numero de apariciones: ')
    diccionario={}
    for x in texto:
        x=x.lower()
        if x.isalpha():
            if x in diccionario:
                diccionario[x]+=1
            else:
                diccionario[x]=1
    return diccionario
funcion()

{'h': 1, 'o': 1, 'l': 2, 'a': 2, 'q': 1, 'u': 1, 'e': 1, 't': 1}

Crea una funció que demani 5 noms de plats i el seu nombre de calories. A mesura que s’introdueixen, guarda aquestes dades en un diccionari i retorna’l. Executa la funció i imprimeix el diccionari resultant.

In [82]:
def funcion():
    diccionario={}
    for x in range(5):
        plato=(input('Introduce el nombre del plato: '))
        calorias=int(input(f'Introduce sus calorias de {plato}: '))
        diccionario[plato]=calorias
    return diccionario
funcion()

{'macarrones': 400, 'pizza': 800, 'sopa': 260, 'arroz': 380, 'croissant': 340}

#### Exercicis de data frames

En aquest exercici repetirem les queries de filtres i agrupacions que vam fer amb SQL, però aquest cop amb Python.

In [83]:
import pandas as pd
data = pd.read_csv("csvs/Country(in).csv", sep=";", encoding="latin-1")
data

Code          Name      Continent                     Region  SurfaceArea  \
0    ABW         Aruba  North America                  Caribbean        193.0   
1    AFG   Afghanistan           Asia  Southern and Central Asia     652090.0   
2    AGO        Angola         Africa             Central Africa    1246700.0   
3    AIA      Anguilla  North America                  Caribbean         96.0   
4    ALB       Albania         Europe            Southern Europe      28748.0   
..   ...           ...            ...                        ...          ...   
234  YEM         Yemen           Asia                Middle East     527968.0   
235  YUG    Yugoslavia         Europe            Southern Europe     102173.0   
236  ZAF  South Africa         Africa            Southern Africa    1221037.0   
237  ZMB        Zambia         Africa             Eastern Africa     752618.0   
238  ZWE      Zimbabwe         Africa             Eastern Africa     390757.0   

     IndepYear  Population  LifeExpectancy       GNP    GNPOld  \
0          NaN      103000            78.4     828.0     793.0   
1       1919.0    22720000            45.9    5976.0       NaN   
2       1975.0    12878000            38.3    6648.0    7984.0   
3          NaN        8000            76.1      63.2       NaN   
4       1912.0     3401200            71.6    3205.0    2500.0   
..         ...         ...             ...       ...       ...   
234     1918.0    18112000            59.8    6041.0    5729.0   
235     1918.0    10640000            72.4   17000.0       NaN   
236     1910.0    40377000            51.1  116729.0  129092.0   
237     1964.0     9169000            37.2    3377.0    3922.0   
238     1980.0    11669000            37.8    5951.0    8670.0   

                 LocalName                                GovernmentForm  \
0                    Aruba  Nonmetropolitan Territory of The Netherlands   
1    Afganistan/Afqanestan                               Islamic Emirate   
2                   Angola                                      Republic   
3                 Anguilla                 Dependent Territory of the UK   
4                Shqipëria                                      Republic   
..                     ...                                           ...   
234               Al-Yaman                                      Republic   
235            Jugoslavija                              Federal Republic   
236           South Africa                                      Republic   
237                 Zambia                                      Republic   
238               Zimbabwe                                      Republic   

                 HeadOfState  Capital Code2,  
0                    Beatrix    129.0    AW,  
1              Mohammad Omar      1.0    AF,  
2    José Eduardo dos Santos     56.0    AO,  
3               Elisabeth II     62.0    AI,  
4             Rexhep Mejdani     34.0    AL,  
..                       ...      ...    ...  
234       Ali Abdallah Salih   1780.0    YE,  
235       Vojislav Koštunica   1792.0    YU,  
236              Thabo Mbeki    716.0    ZA,  
237        Frederick Chiluba   3162.0    ZM,  
238         Robert G. Mugabe   4068.0    ZW,  

[239 rows x 15 columns]

    Realitza els següents exercicis de filtres:

Mostra totes les dades dels països que tenen entre 70 i 80 anys d’esperança de vida (LifeExpectancy).

In [179]:
data[(data['LifeExpectancy'] > 70) & (data['LifeExpectancy'] < 80)]

Code                              Name      Continent           Region  \
0    ABW                             Aruba  North America        Caribbean   
3    AIA                          Anguilla  North America        Caribbean   
4    ALB                           Albania         Europe  Southern Europe   
6    ANT              Netherlands Antilles  North America        Caribbean   
7    ARE              United Arab Emirates           Asia      Middle East   
..   ...                               ...            ...              ...   
226  VCT  Saint Vincent and the Grenadines  North America        Caribbean   
227  VEN                         Venezuela  South America    South America   
228  VGB           Virgin Islands, British  North America        Caribbean   
229  VIR              Virgin Islands, U.S.  North America        Caribbean   
235  YUG                        Yugoslavia         Europe  Southern Europe   

     SurfaceArea  IndepYear  Population  LifeExpectancy      GNP   GNPOld  \
0          193.0        NaN      103000            78.4    828.0    793.0   
3           96.0        NaN        8000            76.1     63.2      NaN   
4        28748.0     1912.0     3401200            71.6   3205.0   2500.0   
6          800.0        NaN      217000            74.7   1941.0      NaN   
7        83600.0     1971.0     2441000            74.1  37966.0  36846.0   
..           ...        ...         ...             ...      ...      ...   
226        388.0     1979.0      114000            72.3    285.0      NaN   
227     912050.0     1811.0    24170000            73.1  95023.0  88434.0   
228        151.0        NaN       21000            75.4    612.0    573.0   
229        347.0        NaN       93000            78.1      0.0      NaN   
235     102173.0     1918.0    10640000            72.4  17000.0      NaN   

                               LocalName  \
0                                  Aruba   
3                               Anguilla   
4                              Shqipëria   
6                   Nederlandse Antillen   
7     Al-Imarat al-´Arabiya al-Muttahida   
..                                   ...   
226     Saint Vincent and the Grenadines   
227                            Venezuela   
228               British Virgin Islands   
229  Virgin Islands of the United States   
235                          Jugoslavija   

                                   GovernmentForm  \
0    Nonmetropolitan Territory of The Netherlands   
3                   Dependent Territory of the UK   
4                                        Republic   
6    Nonmetropolitan Territory of The Netherlands   
7                              Emirate Federation   
..                                            ...   
226                       Constitutional Monarchy   
227                              Federal Republic   
228                 Dependent Territory of the UK   
229                                  US Territory   
235                              Federal Republic   

                     HeadOfState  Capital Code2,  
0                        Beatrix    129.0    AW,  
3                   Elisabeth II     62.0    AI,  
4                 Rexhep Mejdani     34.0    AL,  
6                        Beatrix     33.0    AN,  
7    Zayid bin Sultan al-Nahayan     65.0    AE,  
..                           ...      ...    ...  
226                 Elisabeth II   3066.0    VC,  
227            Hugo Chávez Frías   3539.0    VE,  
228                 Elisabeth II    537.0     VG  
229               George W. Bush   4067.0     VI  
235           Vojislav Koštunica   1792.0    YU,  

[106 rows x 15 columns]

Mostra els noms dels països que es troben al continent europeu i a la regió nòrdica.

In [89]:
data[(data['Continent'].str.contains('europe', case=False))
     & (data['Region'].str.contains('nordic', case=False))]

Code                    Name Continent            Region  SurfaceArea  \
59   DNK                 Denmark    Europe  Nordic Countries      43094.0   
69   FIN                 Finland    Europe  Nordic Countries     338145.0   
73   FRO           Faroe Islands    Europe  Nordic Countries       1399.0   
104  ISL                 Iceland    Europe  Nordic Countries     103000.0   
159  NOR                  Norway    Europe  Nordic Countries     323877.0   
189  SJM  Svalbard and Jan Mayen    Europe  Nordic Countries      62422.0   
200  SWE                  Sweden    Europe  Nordic Countries     449964.0   

     IndepYear  Population  LifeExpectancy       GNP    GNPOld  \
59       800.0     5330000            76.5  174099.0  169264.0   
69      1917.0     5171300            77.4  121914.0  119833.0   
73         NaN       43000            78.4       0.0       NaN   
104     1944.0      279000            79.4    8255.0    7474.0   
159     1905.0     4478500            78.7  145895.0  153370.0   
189        NaN        3200             NaN       0.0       NaN   
200      836.0     8861400            79.6  226492.0  227757.0   

                 LocalName                 GovernmentForm  \
59                 Danmark        Constitutional Monarchy   
69                   Suomi                       Republic   
73                 Føroyar                Part of Denmark   
104                 Ísland                       Republic   
159                  Norge        Constitutional Monarchy   
189  Svalbard og Jan Mayen  Dependent Territory of Norway   
200                Sverige        Constitutional Monarchy   

                HeadOfState  Capital Code2,  
59             Margrethe II   3315.0    DK,  
69            Tarja Halonen   3236.0    FI,  
73             Margrethe II    901.0    FO,  
104  Ólafur Ragnar Grímsson   1449.0    IS,  
159                Harald V   2807.0    NO,  
189                Harald V    938.0    SJ,  
200         Carl XVI Gustaf   3048.0    SE,

Mostra els 3 països amb més esperança de vida (LifeExpectancy) dels grups de països d’Europa Bàltica i Nord-amèrica Carib.

In [117]:
grupo=data[(data['Region'].str.contains('Baltic', case=False))
     | ((data['Continent'].str.contains('North', case=False))
     & (data['Region'].str.contains('Caribbean', case=False)))]
filtrado=grupo.sort_values(by='LifeExpectancy', ascending=False)
filtrado[['Name','Continent', 'Region', 'LifeExpectancy']].head(3)

Name      Continent     Region  LifeExpectancy
53   Cayman Islands  North America  Caribbean            78.9
0             Aruba  North America  Caribbean            78.4
146      Martinique  North America  Caribbean            78.3

Mostra els països que comencen per la lletra “i” o per la lletra “o”.

In [131]:
for x in data['Name']:
    if x.startswith('I') or x.startswith('O'):
        print(x)

Indonesia
India
Ireland
Iran
Iraq
Iceland
Israel
Italy
Oman


In [180]:
data[data['Name'].str.startswith('I')
     | data['Name'].str.startswith('O')]

Code       Name Continent                     Region  SurfaceArea  \
98   IDN  Indonesia      Asia             Southeast Asia    1904569.0   
99   IND      India      Asia  Southern and Central Asia    3287263.0   
101  IRL    Ireland    Europe            British Islands      70273.0   
102  IRN       Iran      Asia  Southern and Central Asia    1648195.0   
103  IRQ       Iraq      Asia                Middle East     438317.0   
104  ISL    Iceland    Europe           Nordic Countries     103000.0   
105  ISR     Israel      Asia                Middle East      21056.0   
106  ITA      Italy    Europe            Southern Europe     301316.0   
163  OMN       Oman      Asia                Middle East     309500.0   

     IndepYear  Population  LifeExpectancy        GNP     GNPOld  \
98      1945.0   212107000            68.0    84982.0   215002.0   
99      1947.0  1013662000            62.5   447114.0   430572.0   
101     1921.0     3775100            76.8    75921.0    73132.0   
102     1906.0    67702000            69.7   195746.0   160151.0   
103     1932.0    23115000            66.5    11500.0        NaN   
104     1944.0      279000            79.4     8255.0     7474.0   
105     1948.0     6217000            78.6    97477.0    98577.0   
106     1861.0    57680000            79.0  1161755.0  1145372.0   
163     1951.0     2542000            71.8    16904.0    16153.0   

            LocalName        GovernmentForm                    HeadOfState  \
98          Indonesia              Republic              Abdurrahman Wahid   
99       Bharat/India      Federal Republic       Kocheril Raman Narayanan   
101      Ireland/Éire              Republic                  Mary McAleese   
102              Iran      Islamic Republic  Ali Mohammad Khatami-Ardakani   
103          Al-´Iraq              Republic      Saddam Hussein al-Takriti   
104            Ísland              Republic         Ólafur Ragnar Grímsson   
105  Yisra’el/Isra’il              Republic                   Moshe Katzav   
106            Italia              Republic           Carlo Azeglio Ciampi   
163             ´Uman  Monarchy (Sultanate)                Qabus ibn Sa´id   

     Capital Code2,  
98     939.0    ID,  
99    1109.0    IN,  
101   1447.0    IE,  
102   1380.0    IR,  
103   1365.0    IQ,  
104   1449.0    IS,  
105   1450.0    IL,  
106   1464.0    IT,  
163   2821.0    OM,

    Realitza els següents exercicis d'agrupacions:

Mostra l’esperança de vida mitjana (LifeExpectancy) de cada continent i regió.

In [142]:
data.groupby(['Continent', 'Region'])[['LifeExpectancy']].mean()

LifeExpectancy
Continent     Region                                   
Africa        Central Africa                  50.311111
              Eastern Africa                  50.810526
              Northern Africa                 65.385714
              Southern Africa                 44.820000
              Western Africa                  52.741176
Antarctica    Antarctica                            NaN
Asia          Eastern Asia                    75.250000
              Middle East                     70.566667
              Southeast Asia                  64.400000
              Southern and Central Asia       61.350000
Europe        Baltic Countries                69.000000
              British Islands                 77.250000
              Eastern Europe                  69.930000
              Nordic Countries                78.333333
              Southern Europe                 76.528571
              Western Europe                  78.255556
North America Caribbean                       73.058333
              Central America                 71.025000
              North America                   75.820000
Oceania       Australia and New Zealand       78.800000
              Melanesia                       67.140000
              Micronesia                      68.085714
              Micronesia/Caribbean                  NaN
              Polynesia                       70.733333
South America South America                   70.946154

Mostra la població total (Population) de cada continent.

In [144]:
data.groupby('Continent')[['Population']].sum()

Population
Continent                
Africa          784475000
Antarctica              0
Asia           3705025700
Europe          730074600
North America   482993000
Oceania          30401150
South America   345780000

Mostra la superfície més petita (SurfaceArea) de cada regió dins de cada continent.

In [ ]:
grupo=data.groupby(['Continent', 'Region'])[['SurfaceArea']].sum().reset_index()
grupo_ordenado=grupo.sort_values('SurfaceArea').groupby('Continent').first().reset_index()
print(grupo_ordenado)

       Continent                Region  SurfaceArea
0         Africa       Southern Africa    2674778.0
1     Antarctica            Antarctica   13132101.0
2           Asia        Southeast Asia    4494801.0
3         Europe      Baltic Countries     175117.0
4  North America             Caribbean     234423.0
5        Oceania  Micronesia/Caribbean         16.0
6  South America         South America   17864926.0


In [185]:
data.groupby(['Continent','Region'])['SurfaceArea'].min()

Continent      Region                   
Africa         Central Africa                  964.0
               Eastern Africa                   78.0
               Northern Africa              163610.0
               Southern Africa               17364.0
               Western Africa                  314.0
Antarctica     Antarctica                       59.0
Asia           Eastern Asia                     18.0
               Middle East                     694.0
               Southeast Asia                  618.0
               Southern and Central Asia       298.0
Europe         Baltic Countries              45227.0
               British Islands               70273.0
               Eastern Europe                33851.0
               Nordic Countries               1399.0
               Southern Europe                   0.4
               Western Europe                    1.5
North America  Caribbean                        96.0
               Central America               21041.0
     

Mostra quants països té cada continent.

In [172]:
data.groupby('Continent')[['Name']].count()

Name
Continent          
Africa           58
Antarctica        5
Asia             51
Europe           46
North America    37
Oceania          28
South America    14

Mostra els continents que tenen més de 50 països.

In [176]:
paises=data.groupby('Continent')[['Name']].count()
top=paises[paises['Name']>50]
print(top)

           Name
Continent      
Africa       58
Asia         51


Pots comprovar les solucions en aquest fitxer: ->python_dataframe_country.ipynb


Pots fer el mateix que has fet en la tasca M7 T13 sobre la taula Country, amb les taules City i Countrylanguage. Aquí tens els fitxers csv. -> City.csv ->Countrylanguage.csv


Pots revisar l'enunciat dels exercicis en les tasques M3 T2 i M3 T3.

El resultat el pots comprovar amb el resultat de les queries de SQL que fes en aquests exercicis

In [ ]:
#Chuleta
.isna() #detecta nulos y devuelve True/False
.isalpha() #detecta el caracter y devuelve True si es una letra, no incluiríaa espacios como len
.lower() # convertir a minusculas
.replace("lo que queremos quitar","lo que reemplaza")
.split() #divide una frase en una lista de palabras, se puede almacenar en una variable.
set() #para crear un conjunto, los conjuntos solo pueden tener valores unicos.
#El numero es par? o impar?
int(num)
if num % 2 == 0:
    return True or False
#agrupa x segun su y, calcula algo y agrupado por x:
df.groupby('x')[['y']]
df.groupby(['x','x2'])[['y']]#hay mas de variable por la que agrupar
df.groupby() el resultado tiene el índice compuesto por las columnas que agrupaste. ==>.reset_index(): baja esas partes del índice a columnas normales otra vez como en un excel (podemos acceder a ellas, agruparlas, ordenarlas, etc)
#PISTA:<pandas.core.groupby.generic.DataFrameGroupBy object at 0x10f92f1d0>, una agrupación preparada, pero aún no agregada, fala una función como sum(), mean(), etc
#groupby() no devuelve un DataFrame directamente, no puedes usar sort_values() directamente. Hay que usar un sum() o otra cosa entremedias
#Clasifica en un grupo o mas de un filtro a la vez:
grupo=df[(df['Columna'].str.contains('Algo', case=False)) #mejor si algo puede aparecer de diferentes formas
   | (df['Columna']=='ALgo')
   & (df['Columna']>60)]
print(len(grupo2))
#Obten los 3 elementos con mas tamaño
df_filtrada= df.sort_values(by='Tamaño', ascending=False)#Por defecto es en orden ascendente, de menos a mas
return df_filtrada[['Name', 'Tamaño']].head(3)

hello///how are you


In [ ]:
#Comienza o acaba con una letra
variable.startswith('')
variable.endswith('')
df[''].str.startswith('')

In [ ]:
import warnings
 
# Para ignorar todos los warnings
warnings.filterwarnings("ignore")